lib/Sema/CSDiagnostics.cpp

```diff
@@ -3493,3 +3493,62 @@ bool MutatingMemberRefOnImmutableBase::diagnoseAsError() {
                             diagIDmember);
   return failure.diagnoseAsError();
 }
+
+bool InvalidTupleSplatWithSingleParameterFailure::diagnoseAsError() {
+  auto *anchor = getRawAnchor();
+
+  auto selectedOverload = getChoiceFor(anchor);
+  if (!selectedOverload || !selectedOverload->choice.isDecl())
+    return false;
+
+  auto *choice = selectedOverload->choice.getDecl();
+
+  auto *argExpr = getArgumentExprFor(anchor);
+  if (!argExpr)
+    return false;
+
+  using Substitution = std::pair<GenericTypeParamType *, Type>;
+  llvm::SetVector<Substitution> substitutions;
+  auto paramTy = ParamType.transform([&](Type type) -> Type {
+    if (auto *typeVar = type->getAs<TypeVariableType>()) {
+      type = resolveType(typeVar);
+
+      if (auto *GP = typeVar->getImpl().getGenericParameter()) {
+        substitutions.insert(std::make_pair(GP, type));
+        return GP;
+      }
+    }
+
+    return type;
+  });
+
+  DeclBaseName name = choice->getBaseName();
+
+  std::string subsStr;
+  if (!substitutions.empty()) {
+    subsStr += " [where ";
+    interleave(
+        substitutions,
+        [&subsStr](const Substitution &substitution) {
```

Comment:
A few comments here:

1) Should we use `[with ` instead, because `where` is a keyword and you're not really describing a `where` clause
2) How about sorting substitutions based on depth/index so we get a consistent ordering no matter how the type is spelled?
3) Can you factor this "compute the substitution string for a given type" into a separate routine? It's awesome and I think we're going to be using it a lot.

Reply:
1. Absolutely, I think you mentioned `with` but for some reason I still went with `where`...
2. Is there a data structure which could do this already or should I roll my own ordering for a vector/set?
3. Sure, I can do that!

Reply:
For 2. , you can sort the vector using the ordering established by `GenericParamKey`.

Reply:
Alright, all done.

1. Switched the message over to `with` instead of `where`.
2. Switched over to `DenseMap<GenericParamKey, std::pair<GenericTypeParamType *, Type>>`
3. Extracted logic into `restoreGenericParameters` with an optional callback.

Reply:
Thanks! Looks great